test/forms/mappings/ConstraintsSpec.scala

```diff
@@ -0,0 +1,191 @@
+      val result = firstError(maxLength(10, "error.length"), regexp("""^\w+$""", "error.regexp"))("a" * 11)
+      result mustEqual Invalid("error.length", 10)
+    }
+
+    "must return Invalid when the second constraint fails" in {
+      val result = firstError(maxLength(10, "error.length"), regexp("""^\w+$""", "error.regexp"))("")
+      result mustEqual Invalid("error.regexp", """^\w+$""")
+    }
+
+    "must return Invalid for the first error when both constraints fail" in {
+      val result = firstError(maxLength(-1, "error.length"), regexp("""^\w+$""", "error.regexp"))("")
+      result mustEqual Invalid("error.length", -1)
+    }
+  }
+
+  "minimumValue" - {
+
+    "must return Valid for a number greater than the threshold" in {
+      val result = minimumValue(1, "error.min").apply(2)
+      result mustEqual Valid
+    }
+
+    "must return Valid for a number equal to the threshold" in {
+      val result = minimumValue(1, "error.min").apply(1)
+      result mustEqual Valid
+    }
+
+    "must return Invalid for a number below the threshold" in {
+      val result = minimumValue(1, "error.min").apply(0)
+      result mustEqual Invalid("error.min", 1)
+    }
+  }
+
+  "maximumValue" - {
+
+    "must return Valid for a number less than the threshold" in {
+      val result = maximumValue(1, "error.max").apply(0)
+      result mustEqual Valid
+    }
+
+    "must return Valid for a number equal to the threshold" in {
+      val result = maximumValue(1, "error.max").apply(1)
+      result mustEqual Valid
+    }
+
+    "must return Invalid for a number above the threshold" in {
+      val result = maximumValue(1, "error.max").apply(2)
+      result mustEqual Invalid("error.max", 1)
+    }
+  }
+
+  "regexp" - {
+
+    "must return Valid for an input that matches the expression" in {
+      val result = regexp("""^\w+$""", "error.invalid")("foo")
+      result mustEqual Valid
+    }
+
+    "must return Invalid for an input that does not match the expression" in {
+      val result = regexp("""^\d+$""", "error.invalid")("foo")
+      result mustEqual Invalid("error.invalid", """^\d+$""")
+    }
+  }
+
+  "maxLength" - {
+
+    "must return Valid for a string shorter than the allowed length" in {
+      val result = maxLength(10, "error.length")("a" * 9)
+      result mustEqual Valid
+    }
+
+    "must return Valid for an empty string" in {
+      val result = maxLength(10, "error.length")("")
+      result mustEqual Valid
+    }
+
+    "must return Valid for a string equal to the allowed length" in {
+      val result = maxLength(10, "error.length")("a" * 10)
+      result mustEqual Valid
+    }
+
+    "must return Invalid for a string longer than the allowed length" in {
+      val result = maxLength(10, "error.length")("a" * 11)
+      result mustEqual Invalid("error.length", 10)
+    }
+  }
+
+  "maxDate" - {
+
+    "must return Valid for a date before or equal to the maximum" in {
+
+      val gen: Gen[(LocalDate, LocalDate)] = for {
```

Comment:
this date generator seems to be duplicated across this spec, would be nice to move it under main generator